### PR TITLE
feat: Rebase `feature/optional` to `develop`

### LINKED ISSUE #1749: [C++17] Allow std::optional to convert to nlohmann::json
This is what I assume to be a fairly minor change to the library's logic, but would represent a large improvement in usability.

## Feature request

C++17 adds `std::optional<T>`. Let `std::optional<T>` convert to `nlohmann::json`.

If `std::optional<T>` is empty, I would expect this to result in `null` - if `std::optional<T>`, I would expect it to result in the contained value converted.

## Use case example

The change would simplify (especially nested) uses of `std::optional`.

For example, imagine a simple key-value map, where each integer key can optionally have a localized string representation for each of 8 locales. The client dynamically selects the "best" available locale depending on user preferences.

In server code, one could express this as a `std::vector<std::array<std::optional<std::string>, 8>>`.
Currently, this type does not convert to `nlohmann::json` because `std::optional<std::string>` does not - resulting in needing to allocate a bunch of temporaries so the inner conversions can be done "manually".

Allowing `std::optional` to convert to `nlohmann::json` would simplify this logic greatly.

### PR DESCRIPTION
**Pull request by fsandhei**
Rebasing `feature/optional` in attempt to revive this PR, referring to https://github.com/nlohmann/json/pull/2117.

Sorry for long time between responses.

I may have been going forward with this the "wrong" way, so I apologize in advance.

Please let me know if this should be handled differently.

## TODO
There is one test that is failing: Conversions from a default initialized `nlohmann::json` to `std::optional<T>` seems to fail. In the test case below, the last assertion fails with an exception. This is tested on Arch Linux with GCC 13.1.1.

I was not able to figure out how to fix this so I need some assistance here.

```
/home/fredrik/dev/github/cpp/fredriksd/json/tests/src/unit-conversions.cpp:1573:
TEST CASE:  std::optional
  null

/home/fredrik/dev/github/cpp/fredriksd/json/tests/src/unit-conversions.cpp:1573: ERROR: test case THREW exception: exception thrown in subcase - will translate later when the whole test case has been exited (cannot translate while there is an active exception)

===============================================================================
/home/fredrik/dev/github/cpp/fredriksd/json/tests/src/unit-conversions.cpp:1573:
TEST CASE:  std::optional

DEEPEST SUBCASE STACK REACHED (DIFFERENT FROM THE CURRENT ONE):
  null

/home/fredrik/dev/github/cpp/fredriksd/json/tests/src/unit-conversions.cpp:1573: ERROR: test case THREW exception: [json.exception.type_error.302] type must be string, but is null
```

### CODE REVIEW

**Comment by fredriksd** [comment]
Think something got messed up on comparison, so closing to not make noise. Sorry about that.

**Comment by fsandhei** [comment]
Changed the base branch to `develop`.

I'm sorry if I'm making it a bit difficult here; I forked the repository and rebased `feature/optional` to `develop` and created this PR against `nlohmann:feature/develop` but I noticed that included 590 commits (which makes sense) but it cluttered the actual change here, so I instead changed it towards `nlohmann:develop`.

**Comment by lightyear15** [comment]
I am also interested in having ``std::optional`` support landing in future versions of this library.
I see there was a lot of discussion on how to properly support and convert c++ optional fields vs json nullable fields.
Is there a final decision on the matter? How can we revive this PR and push forward support of ``std::optional`` fields?
@nlohmann

File: include/nlohmann/detail/conversions/from_json.hpp
Comment (lightyear15) [inline include/nlohmann/detail/conversions/from_json.hpp]:
```suggestion
        opt = std::make_optional(j.template get<T>());
```

File: tests/src/unit-conversions.cpp
Comment (lightyear15) [inline tests/src/unit-conversions.cpp:1613]:
```suggestion
        CHECK(j_null.get<std::optional<std::string>>() == opt_string);
```

**Comment by fsandhei** [comment]
See that clang-tidy is failing on some checks. Apparently this is on develop, too. I see that it is related to a supposed new check from clang-tidy v19: https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-designated-initializers.html.

**Comment by nlohmann** [comment]
> See that clang-tidy is failing on some checks. Apparently this is on develop, too. I see that it is related to a supposed new check from clang-tidy v19: [clang.llvm.org/extra/clang-tidy/checks/modernize/use-designated-initializers.html](https://clang.llvm.org/extra/clang-tidy/checks/modernize/use-designated-initializers.html).

I'm on it. May take some more time.

**Comment by lightyear15** [comment]
has this PR stalled again?

**Comment by fsandhei** [comment]
I'm currently blocked by https://github.com/nlohmann/json/pull/4311.

**Comment by fsandhei** [comment]
There has not been any activity in #4311 in a while. I wonder if this would be okay to merge?

**Comment by nlohmann** [comment]
Thanks for keeping this alive. I will have a look at this on Saturday.

**Comment by fsandhei** [comment]
> Thanks for keeping this alive. I will have a look at this on Saturday.

Happy to help! Hopefully this can go in soon.

**Comment by nlohmann** [comment]
Please suppress the last clang-tidy warning and not change the code.

**Comment by fsandhei** [comment]
> Please suppress the last clang-tidy warning and not change the code.

Reverted change with [296549c](https://github.com/nlohmann/json/pull/4036/commits/296549c3ad3941ccf007db543c8e0d3f908a67d0).

**Review by nlohmann — APPROVED** [review]
Looks good to me.

**Comment by nlohmann** [comment]
Thanks!

**Pull request merged**
